cape_privacy/spark/transformations/tokenizer.py

```diff
@@ -60,3 +60,84 @@ def _to_token(self, x: str):
     @property
     def key(self):
         return self._key
+
+
+class ReversibleTokenizer(base.Transformation):
+    """ReversibleTokenizer: map a string to a token to obfuscate it.
+
+    When applying the Tokenizer to a Spark series of type string,
+    each value gets mapped to a token (hexadecimal string).
+    If a value is repeated several times across the series, it always
+    get mapped to the same token in order to maintain the count.
+    A value can be mapped to different tokens by setting the key to a
+    different value.
+
+    This tokenizer allows tokens to be reversed to their original data
+    when the secret key is known.
+
+    Attributes:
+        key: expect a string or byte string.
+    """
+
+    identifier = "reversible-tokenizer"
+    type_signature = "col->col"
+
+    def __init__(self, key):
+        typecheck.check_arg(key, (str, bytes))
+        super().__init__(dtype=dtypes.String)
+        if isinstance(key, str):
+            key = key.encode()
+        if not key or len(key) != 32:
+            raise ValueError("Key must be exactly 32 bytes")
+        self.key = key
+
+    def __call__(self, x):
+        return _series_to_token.map(self.token)
+
+    @functions.pandas_udf(dtypes.String, functions.PandasUDFType.SCALAR)
+    def token(series: pd.Series):
+        return series.map(self._item_to_token)
+
+    def _to_token(self, x: str):
+        cipher = AES.new(key=self.key, mode=AES.MODE_SIV)
+        ciphertext, tag = cipher.encrypt_and_digest(x.encode())
+        assert len(tag) == 16, len(tag)
+        token = tag.hex() + ciphertext.hex()
+        return token
+
+
+class TokenReverser(base.Transformation):
+    """TokenReverser: recover string from token.
+
+    When applying the TokenReverser to a Spark series of tokens,
+    each token is mapped back to the string that was originally used
+    by ReversibleTokenizer to construct the token. The same key must
+    be used.
+
+    Attributes:
```

Comment:
any reason the `encoding` attribute is missing here?

Reply:
good idea!